### PR TITLE
[BEAM-13560][Playground] Split builder into several builders for each step of the pipeline execution

### PR DESCRIPTION
**Pull request by daria-malkova**
[BEAM-13560][Playground]  Split builder into several builders for each step of pipeline execution

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/backend/internal/code_processing/code_processing.go [hunk 1]
```diff
@@ -61,117 +60,65 @@ func Process(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycl
...
-		if sdkEnv.ApacheBeamSdk == pb.Sdk_SDK_JAVA {
-			executor = executorBuilder.WithCompiler().
-				WithFileName(builder.GetFileNameFromFolder(lc.GetAbsoluteSourceFolderPath())).Build() // Need changed name for unit tests
-		}
-		logger.Infof("%s: Compile() ...\n", pipelineId)
-		compileCmd := executor.Compile(pipelineLifeCycleCtx)
-		var compileError bytes.Buffer
-		var compileOutput bytes.Buffer
-		runCmdWithOutput(compileCmd, &compileOutput, &compileError, successChannel, errorChannel)
+func RunStep(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycle, pipelineId uuid.UUID, isUnitTest bool, sdkEnv *environment.BeamEnvs, pipelineOptions string, pipelineLifeCycleCtx context.Context, cancelChannel chan bool) {
```
Comment (AydarZaynutdinov) [inline playground/backend/internal/code_processing/code_processing.go:92]: Why does this method name start with a capital letter?

File: playground/backend/internal/code_processing/code_processing.go [hunk 2]
```diff
@@ -210,30 +157,121 @@ func Process(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycl
...
+	}
+	return executor, false
+}
+
+func PrepareStep(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycle, pipelineId uuid.UUID, sdkEnv *environment.BeamEnvs, pipelineLifeCycleCtx context.Context, validationResults *sync.Map, cancelChannel chan bool) (executors.Executor, bool) {
+	errorChannel, successChannel := createStatusChannels()
+	executorBuilder, err := builder.SetupPreparatorBuilder(lc, sdkEnv)
+	if err != nil {
+		_ = processSetupError(err, pipelineId, cacheService, pipelineLifeCycleCtx)
+		return executors.Executor{}, false
```
[the remark this answers is not in the document]
Comment (daria-malkova) [reply playground/backend/internal/code_processing/code_processing.go:199]: Didn't notice, you are right, fixed.

File: playground/backend/internal/code_processing/code_processing.go [hunk 3]
```diff
@@ -210,30 +157,121 @@ func Process(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycl
...
+	}
+	return executor, false
+}
+
+func ValidateStep(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycle, pipelineId uuid.UUID, sdkEnv *environment.BeamEnvs, pipelineLifeCycleCtx context.Context, validationResults *sync.Map, cancelChannel chan bool) (executors.Executor, bool) {
+	errorChannel, successChannel := createStatusChannels()
+	executorBuilder, err := builder.SetupValidatorBuilder(lc, sdkEnv)
+	if err != nil {
+		_ = processSetupError(err, pipelineId, cacheService, pipelineLifeCycleCtx)
+		return executors.Executor{}, false
```
[the remark this answers is not in the document]
Comment (daria-malkova) [reply playground/backend/internal/code_processing/code_processing.go:231]: Done

File: playground/backend/internal/fs_tool/java_fs.go [hunk 4]
```diff
@@ -36,8 +36,7 @@ func newJavaLifeCycle(pipelineId uuid.UUID, pipelinesFolder string) *LifeCycle {
 }
 
 // executableName returns name that should be executed (HelloWorld for HelloWorld.class for java SDK)
-func executableName(pipelineId uuid.UUID, pipelinesFolder string) (string, error) {
-	baseFileFolder := filepath.Join(pipelinesFolder, pipelineId.String())
+func executableName(baseFileFolder string) (string, error) {
 	binFileFolder := filepath.Join(baseFileFolder, compiledFolderName)
 	dirEntries, err := os.ReadDir(binFileFolder)
```
[the remark this answers is not in the document]
Comment (daria-malkova) [reply playground/backend/internal/fs_tool/java_fs.go:41]: Good suggestion, let's fix it later when DTO will be merged

**Comment by daria-malkova** [comment]
@KhaninArtur @ilya-kozyrev

File: playground/backend/internal/setup_tools/builder/setup_builder.go [hunk 5]
```diff
@@ -31,73 +31,134 @@ const (
...
+func SetupRunBuilder(lc *fs_tool.LifeCycle, pipelineOptions string, sdkEnv *environment.BeamEnvs) (*executors.ExecutorBuilder, error) {
+	sdk := sdkEnv.ApacheBeamSdk
+
+	if sdk == pb.Sdk_SDK_JAVA {
+		pipelineOptions = utils.ReplaceSpacesWithEquals(pipelineOptions)
+	}
+	executorConfig := sdkEnv.ExecutorConfig
+	builder := executors.NewExecutorBuilder().
 		WithRunner().
+		WithExecutableFileName(lc.GetAbsoluteExecutableFilePath()).
```
Comment (KhaninArtur) [inline playground/backend/internal/setup_tools/builder/setup_builder.go:93]: Shouldn't we move this to the `pb.Sdk_SDK_PYTHON` case to keep consistency?

File: playground/backend/internal/setup_tools/builder/setup_builder_test.go [hunk 6]
```diff
@@ -23,100 +23,293 @@ import (
...
+		lc             *fs_tool.LifeCycle
+		executorConfig *environment.ExecutorConfig
+	}
+	tests := []struct {
+		name string
+		args args
+		want []string
+	}{
+		{
+			name: "",
```
Comment (KhaninArtur) [inline playground/backend/internal/setup_tools/builder/setup_builder_test.go:301]: Could you please give a name for this test?

File: playground/backend/internal/code_processing/code_processing.go [hunk 7]
```diff
@@ -210,30 +154,122 @@ func Process(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycl
...
+
+func validateStep(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycle, pipelineId uuid.UUID, sdkEnv *environment.BeamEnvs, pipelineLifeCycleCtx context.Context, validationResults *sync.Map, cancelChannel chan bool) *executors.Executor {
+	errorChannel, successChannel := createStatusChannels()
+	executorBuilder, err := builder.SetupValidatorBuilder(lc, sdkEnv)
+	if err != nil {
+		_ = processSetupError(err, pipelineId, cacheService, pipelineLifeCycleCtx)
+		return nil
+	}
+	executor := executorBuilder.Build()
+	// Validate
```
Comment (KhaninArtur) [inline playground/backend/internal/code_processing/code_processing.go:232]: This comment is "extremely useful", let's remove it😁

File: playground/backend/internal/code_processing/code_processing.go [hunk 8]
```diff
@@ -210,30 +154,122 @@ func Process(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycl
...
+	logger.Infof("%s: Validate() ...\n", pipelineId)
+	validateFunc := executor.Validate()
+	// Run validate function
```
Comment (KhaninArtur) [inline playground/backend/internal/code_processing/code_processing.go:235]: ditto

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 9]
```diff
@@ -532,64 +526,7 @@ func TestGetLastIndex(t *testing.T) {
 	}
 }
 
-func Test_setJavaExecutableFile(t *testing.T) {
```
Comment (KhaninArtur) [inline playground/backend/internal/code_processing/code_processing_test.go:535]: I believe, we also need to test all introduced functions like `validateStep`, `compileStep`, etc.

**Review by KhaninArtur — CHANGES_REQUESTED** [review]
Let's fix the comments, the rest is LGTM!

**Review by AydarZaynutdinov — APPROVED** [review]
LGTM

File: playground/backend/internal/code_processing/code_processing.go [hunk 10]
```diff
@@ -210,30 +154,122 @@ func Process(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycl
...
+
+func prepareStep(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycle, pipelineId uuid.UUID, sdkEnv *environment.BeamEnvs, pipelineLifeCycleCtx context.Context, validationResults *sync.Map, cancelChannel chan bool) *executors.Executor {
+	errorChannel, successChannel := createStatusChannels()
+	executorBuilder, err := builder.SetupPreparatorBuilder(lc, sdkEnv)
+	if err != nil {
+		_ = processSetupError(err, pipelineId, cacheService, pipelineLifeCycleCtx)
+		return nil
+	}
+	executor := executorBuilder.Build()
+	// Prepare
```
Comment (ilya-kozyrev) [inline playground/backend/internal/code_processing/code_processing.go:200]: do we need this comment?

File: playground/backend/internal/setup_tools/builder/setup_builder.go [hunk 11]
```diff
@@ -31,73 +31,134 @@ const (
...
-		WithValidator().
-		WithSdkValidators(val).
 		WithPreparator().
 		WithSdkPreparators(prep).
+		ExecutorBuilder
+	return &builder, err
+}
+
+// SetupCompilerBuilder return executor with set args for compiler
+func SetupCompilerBuilder(lc *fs_tool.LifeCycle, sdkEnv *environment.BeamEnvs) *executors.ExecutorBuilder {
```
Comment (ilya-kozyrev) [inline playground/backend/internal/setup_tools/builder/setup_builder.go:63]: What about naming it just Compiler? Then we could use it like that:  `setup_tools.builder.Compiler()`
and the same for all other builders in setup_tools

**Review by ilya-kozyrev — CHANGES_REQUESTED** [review]

File: playground/backend/internal/code_processing/code_processing.go (hunk 10 above)
Comment (daria-malkova) [reply playground/backend/internal/code_processing/code_processing.go]: No, done

File: playground/backend/internal/code_processing/code_processing.go (hunk 7 above)
Comment (daria-malkova) [reply playground/backend/internal/code_processing/code_processing.go]: Done

File: playground/backend/internal/code_processing/code_processing.go (hunk 8 above)
Comment (daria-malkova) [reply playground/backend/internal/code_processing/code_processing.go]: Done

File: playground/backend/internal/setup_tools/builder/setup_builder.go (hunk 11 above)
Comment (daria-malkova) [reply playground/backend/internal/setup_tools/builder/setup_builder.go]: Done

File: playground/backend/internal/setup_tools/builder/setup_builder_test.go (hunk 6 above)
Comment (daria-malkova) [reply playground/backend/internal/setup_tools/builder/setup_builder_test.go]: I will :)

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 9 above)
Comment (daria-malkova) [reply playground/backend/internal/code_processing/code_processing_test.go:535]: Done

File: playground/backend/internal/fs_tool/java_fs.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (daria-malkova) [reply playground/backend/internal/fs_tool/java_fs.go]: Done

File: playground/backend/internal/setup_tools/builder/setup_builder.go (hunk 5 above)
Comment (daria-malkova) [reply playground/backend/internal/setup_tools/builder/setup_builder.go]: Done

File: playground/backend/internal/setup_tools/builder/setup_builder_test.go (hunk 6 above)
Comment (daria-malkova) [reply playground/backend/internal/setup_tools/builder/setup_builder_test.go]: Done

File: playground/backend/internal/setup_tools/builder/setup_builder.go [hunk 12]
```diff
@@ -31,73 +31,139 @@ const (
 	javaLogConfigFilePlaceholder = "{logConfigFile}"
 )
 
-// SetupExecutorBuilder return executor with set args for validator, preparator, compiler and runner
-func SetupExecutorBuilder(paths fs_tool.LifeCyclePaths, pipelineOptions string, sdkEnv *environment.BeamEnvs) (*executors.ExecutorBuilder, error) {
+// Validator return executor with set args for validator
+func Validator(paths *fs_tool.LifeCyclePaths, sdkEnv *environment.BeamEnvs) (*executors.ExecutorBuilder, error) {
```
Comment (KhaninArtur) [inline playground/backend/internal/setup_tools/builder/setup_builder.go:35]: Why do you pass here the pointer? I think we don't want to change anything in `paths`, let's remove `*`, what do you think?

File: playground/backend/internal/setup_tools/builder/setup_builder.go [hunk 13]
```diff
@@ -31,73 +31,139 @@ const (
...
 	}
+	builder := executors.NewExecutorBuilder().
+		WithValidator().
+		WithSdkValidators(val).
+		ExecutorBuilder
+	return &builder, err
+}
+
+// Preparer return executor with set args for preparator
+func Preparer(paths *fs_tool.LifeCyclePaths, sdkEnv *environment.BeamEnvs) (*executors.ExecutorBuilder, error) {
```
Comment (KhaninArtur) [inline playground/backend/internal/setup_tools/builder/setup_builder.go:49]: ditto

File: playground/backend/internal/setup_tools/builder/setup_builder.go [hunk 14]
```diff
@@ -31,73 +31,139 @@ const (
...
-		WithValidator().
-		WithSdkValidators(val).
 		WithPreparator().
 		WithSdkPreparators(prep).
+		ExecutorBuilder
+	return &builder, err
+}
+
+// Compiler return executor with set args for compiler
+func Compiler(paths *fs_tool.LifeCyclePaths, sdkEnv *environment.BeamEnvs) *executors.ExecutorBuilder {
```
Comment (KhaninArtur) [inline playground/backend/internal/setup_tools/builder/setup_builder.go:63]: ditto

File: playground/backend/internal/setup_tools/builder/setup_builder.go (hunk 12 above)
Comment (KhaninArtur) [reply playground/backend/internal/setup_tools/builder/setup_builder.go:35]: Maybe, we can also remove `*` from `sdkEnv`?

File: playground/backend/internal/setup_tools/builder/setup_builder.go [hunk 15]
```diff
@@ -31,73 +31,139 @@ const (
...
+	case pb.Sdk_SDK_JAVA:
+		builder.
+			WithCompiler().
+			WithFileName(GetFirstFileFromFolder(paths.AbsoluteSourceFileFolderPath))
+	}
+	return &builder
+}
+
+// Runner return executor with set args for runner
+func Runner(paths *fs_tool.LifeCyclePaths, pipelineOptions string, sdkEnv *environment.BeamEnvs) (*executors.ExecutorBuilder, error) {
```
Comment (KhaninArtur) [inline playground/backend/internal/setup_tools/builder/setup_builder.go:84]: ditto

File: playground/backend/internal/setup_tools/builder/setup_builder.go [hunk 16]
```diff
@@ -31,73 +31,139 @@ const (
...
+	case pb.Sdk_SDK_JAVA: // Executable name for java class is known after compilation
+		args := replaceLogPlaceholder(paths, executorConfig)
+		className, err := paths.ExecutableName(paths.AbsoluteExecutableFileFolderPath)
+		if err != nil {
+			return nil, fmt.Errorf("no executable file name found for JAVA pipeline at %s", paths.AbsoluteExecutableFileFolderPath)
 		}
-		builder = builder.WithRunner().WithArgs(args).ExecutorBuilder
-		builder = builder.WithTestRunner().WithWorkingDir(paths.AbsoluteBaseFolderPath).ExecutorBuilder //change directory for unit test
+		builder = builder.
+			WithRunner().
```
Comment (KhaninArtur) [inline playground/backend/internal/setup_tools/builder/setup_builder.go:107]: We already have `WithRunner()` at the beginning of the building of the executor, should we call it once more here?

File: playground/backend/internal/setup_tools/builder/setup_builder.go [hunk 17]
```diff
@@ -31,73 +31,139 @@ const (
...
+			WithArgs(args).
+			WithExecutableFileName(className).
+			ExecutorBuilder
 	case pb.Sdk_SDK_GO: //go run command is executable file itself
 		builder = builder.
-			WithExecutableFileName("").
 			WithRunner().
```
Comment (KhaninArtur) [inline playground/backend/internal/setup_tools/builder/setup_builder.go:113]: ditto

File: playground/backend/internal/setup_tools/builder/setup_builder.go [hunk 18]
```diff
@@ -31,73 +31,139 @@ const (
...
+			WithCommand(paths.AbsoluteExecutableFilePath).
+			ExecutorBuilder
 	case pb.Sdk_SDK_PYTHON:
-		builder = *builder.WithExecutableFileName(paths.AbsoluteExecutableFilePath)
-	case pb.Sdk_SDK_SCIO:
-		return nil, fmt.Errorf("SCIO is not supported yet")
-	default:
-		return nil, fmt.Errorf("incorrect sdk: %s", sdkEnv.ApacheBeamSdk)
+		builder = builder.
+			WithRunner().
```
Comment (KhaninArtur) [inline playground/backend/internal/setup_tools/builder/setup_builder.go:119]: ditto

File: playground/backend/internal/setup_tools/builder/setup_builder.go [hunk 19]
```diff
@@ -31,73 +31,139 @@ const (
...
+			WithExecutableFileName(paths.AbsoluteExecutableFilePath).
+			ExecutorBuilder
+	}
+	return &builder, nil
+}
+
+// TestRunner return executor with set args for runner
+func TestRunner(paths *fs_tool.LifeCyclePaths, sdkEnv *environment.BeamEnvs) (*executors.ExecutorBuilder, error) {
```
Comment (KhaninArtur) [inline playground/backend/internal/setup_tools/builder/setup_builder.go:127]: ditto

File: playground/backend/internal/setup_tools/builder/setup_builder.go [hunk 20]
```diff
@@ -31,73 +31,139 @@ const (
...
+			ExecutorBuilder //change directory for unit test
 	}
 	return &builder, nil
 }
 
-// GetFileNameFromFolder return a name of the first file in a specified folder
-func GetFileNameFromFolder(folderAbsolutePath, extension string) string {
-	files, _ := filepath.Glob(fmt.Sprintf("%s/*%s", folderAbsolutePath, extension))
+// replaceLogPlaceholder replaces placeholder for log for JAVA SDK
+func replaceLogPlaceholder(paths *fs_tool.LifeCyclePaths, executorConfig *environment.ExecutorConfig) []string {
```
Comment (KhaninArtur) [inline playground/backend/internal/setup_tools/builder/setup_builder.go:153]: ditto

**Review by KhaninArtur — CHANGES_REQUESTED** [review]

File: playground/backend/internal/setup_tools/builder/setup_builder.go (hunk 16 above)
Comment (daria-malkova) [reply playground/backend/internal/setup_tools/builder/setup_builder.go:107]: Yes

File: playground/backend/internal/setup_tools/builder/setup_builder.go (hunk 17 above)
Comment (daria-malkova) [reply playground/backend/internal/setup_tools/builder/setup_builder.go:113]: we need it

File: playground/backend/internal/setup_tools/builder/setup_builder.go (hunk 12 above)
Comment (daria-malkova) [reply playground/backend/internal/setup_tools/builder/setup_builder.go:35]: Let's do it at the next PR, it is not very critical right now

File: playground/backend/internal/setup_tools/builder/setup_builder.go (hunk 13 above)
Comment (daria-malkova) [reply playground/backend/internal/setup_tools/builder/setup_builder.go:49]: same

File: playground/backend/internal/setup_tools/builder/setup_builder.go (hunk 14 above)
Comment (daria-malkova) [reply playground/backend/internal/setup_tools/builder/setup_builder.go:63]: same

File: playground/backend/internal/setup_tools/builder/setup_builder.go (hunk 18 above)
Comment (daria-malkova) [reply playground/backend/internal/setup_tools/builder/setup_builder.go:119]: same

File: playground/backend/internal/setup_tools/builder/setup_builder.go (hunk 20 above)
Comment (daria-malkova) [reply playground/backend/internal/setup_tools/builder/setup_builder.go:153]: same

File: playground/backend/internal/setup_tools/builder/setup_builder.go (hunk 19 above)
Comment (daria-malkova) [reply playground/backend/internal/setup_tools/builder/setup_builder.go:127]: same

File: playground/backend/internal/setup_tools/builder/setup_builder.go (hunk 15 above)
Comment (daria-malkova) [reply playground/backend/internal/setup_tools/builder/setup_builder.go:84]: next PR

**Comment by daria-malkova** [comment]
R: @pabloem

**Review by KhaninArtur — APPROVED** [review]
LGTM

**Comment by pabloem** [comment]
ah this LGTM. Thanks all!

**Pull request merged**